### ISSUE TITLE
Two models for embedding in parallel

### ISSUE THREAD

**Issue by akset2X** [neuml/txtai#524]
I am using txtAI functionality through REST API. Currently, I'm using all-MiniLM-L6-v2 but can I have another model for embedding at the same time ?

At the upsert/index time and search time, I should have the option to choose which model to use. That I will be choosing based on the language of the text. How can I achieve this?

**Comment by davidmezzetti** [comment]
With 6.0, you can now have multiple indexes. See the example below for applications.

```python
# Create embeddings with subindexes
embeddings = Embeddings(
  content=True,
  defaults=False,
  indexes={
    "index1": {
      "path": "sentence-transformers/all-MiniLM-L6-v2"
    },
    "index2": {
      "path": "sentence-transformers/nli-mpnet-base-v2"
    }
  }
)
embeddings.index(data)

# Query index1
embeddings.search("select id, text, score from txtai where similar('query', 'index1')")

# Query index2
embeddings.search("select id, text, score from txtai where similar('query', 'index2')")
```

The same can be done through an application/API. The config would be.

```yaml
embeddings:
  content: True
  defaults: False
  indexes: 
    index1:
      path: "sentence-transformers/all-MiniLM-L6-v2"
    index2:
      path: "sentence-transformers/nli-mpnet-base-v2"
```

Now this builds two separate indexes both with all the data. If you're looking to have index1 and index2 have a subset of the data, I'd create two separate fields in your JSON. For example:

```python
document = {"text1: "text1", "text2", "text2"}
```

And config to set the index fields.

```yaml
embeddings:
  content: True
  defaults: False
  indexes: 
    index1:
      path: "sentence-transformers/all-MiniLM-L6-v2"
      columns:
        text: text1
    index2:
      path: "sentence-transformers/nli-mpnet-base-v2"
      columns:
        text: text2
```

You can read more on this in this article: https://medium.com/neuml/whats-new-in-txtai-6-0-7d93eeedf804

**Comment by akset2X** [comment]
Oh thanks! I will give this a try.

**Comment by akset2X** [comment]
This works in colab, both API and prompt approach. But when I tried the same in my local, for some reason I observe an unexpected behavior.

In colab, when I specified the two models in yml and started the API instance, after completing upsert operation I could see that two index directories getting created with unique model name inside each folder.

But in my local only one index directory is getting created and when I read the config file I am seeing two model names in it.

**index.yml**
![Screenshot (341)](https://github.com/neuml/txtai/assets/117704649/cfb3beec-a280-4ab7-b84f-5f93d6087ab1)


**config**
![Screenshot (340)](https://github.com/neuml/txtai/assets/117704649/a706274e-c795-40c9-93e2-d45b6b6b3579)

As a result the search is not working as expected, it is always giving same results with same score irrespective of the index name that I give inside the similar() call of the SELECT query.

**Comment by akset2X** [comment]
No, I am just seeing config, embeddings, and document those 3 files. I don't see any directories in the name of index1 and index2. If you check my screenshot above the path I specified is ./tmp/indexes so inside that I just have three files alone.

Regarding my disk storage I have 329GB free space.

**Comment by davidmezzetti** [comment]
I don't really have any ideas on what is going on. I'd recommend comparing the environment that is working and the one that isn't and see if there is a subtle difference.

The way you describe the files saved doesn't match the config. For example, there shouldn't be an embeddings file since defaults is set to false.

**Comment by akset2X** [comment]
I somehow managed to make it work. I was using Python v3.9 and then I upgraded it to v3.10. May be that solved the issue and now, it works as expected. 

I have one more thing to ask currently using this approach we are storing the embeddings in both the index right, is there anyway I can only upsert to index1 or index2 whichever I choose? This option is allowed at search time. Can we have this option at index/upsert time also ?

**Comment by davidmezzetti** [comment]
Glad to hear it worked.

If you want to conditionally index data, the best thing to do is to set the index column name. For example:

```python
    indexes={
        "index1": {
            "columns": {
                "text": "column1"
            },
        },
        "index2":{
            "columns": {
                "text": "column2"
            }
        }
    }
```

Then you can conditionally populate that field using whatever logic you choose.

**Comment by davidmezzetti** [comment]
Closing this due to inactivity. Please re-open or open a new issue to continue the discussion.

**Comment by akset2X** [comment]
```
indexes={
"index1": {
  "path": "sentence-transformers/all-MiniLM-L6-v2"
  "columns": {
    "text": "column1"
  },
},
"index2":{
  "path": "sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2"
  "columns": {
    "text": "column2"
  }
}
}
```
This is good. I understand that with the help of `columns` if I don't want index1 to process a text while indexing/upserting I can make the `"column1"` field empty and just put the text in the field `"column2"`, this way all-MiniLM would not process the text but paraphrase-multilingual will do. Am I right?

Example:
```
data = [
{"id": 0, "column1": "Some english content", "column2": "Some French content"},
{"id": 1, "column1": "Some english content"}
{"id": 2, "column2": "Some French content"}]
```
In this example, I have declared the variable `data`. If I /upsert it,
* data with `id 0` will be trained by both the indexes (all-MiniLM and paraphrase)
* data with `id 1` will be trained by only index1 (all-MiniLM)
* data with `id 2` will be trained by only index2 (paraphrase...)

Did I get it correctly?

**Comment by davidmezzetti** [comment]
That's the right idea. If you run into any issues or problems, let me know.

**Comment by akset2X** [comment]
**Thanks for confirming. It just works fine.**

Actually, to achieve this same usecase, before finding this multi-index(`indexes`) option my initial plan was to use two API instances of txtai running on separate ports. So that both the `index`'s will be completely segregated. 
I even tried to test that approach. I tried as follows,

One of the txtai instances uses the all-MiniLM model, and the other one uses the paraphrase-MiniLM model for embedding. Additionally, each instance uses a different path for index or content writing.

However, that approach did not work as expected because when both instances are running, I encountered an issue where I couldn't upsert texts to the correct index that I intended. Even though the API call went to the right instance, the upsert and search operations were happening on the other one. Also, the upsert operation fails with db lock issue sometime. By any chance if I get to know why this happening it will be great. I guess this is some Python issue related to single thread or something. 

Any idea on this ? It will be helpful for those who wants to use two instances with different config at the same time.
